File: packages/typescript/src/introspect.ts
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { InMemoryTransport } from "@modelcontextprotocol/sdk/inMemory.js";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";

export interface IntrospectionTool {
  name: string;
  title?: string;
  description?: string;
  annotations?: Record<string, unknown>;
  inputSchema?: Record<string, unknown>;
  outputSchema?: Record<string, unknown>;
}

export interface IntrospectionResource {
  uri: string;
  name?: string;
  description?: string;
  mimeType?: string;
}

export interface IntrospectionPrompt {
  name: string;
  description?: string;
  arguments?: IntrospectionPromptArgument[];
}

export interface IntrospectionPromptArgument {
  name: string;
  description?: string;
  required?: boolean;
}

export interface IntrospectionResult {
  tools: IntrospectionTool[];
  resources: IntrospectionResource[];
  prompts: IntrospectionPrompt[];
  capabilities: Record<string, Record<string, unknown> | object>;
}

export async function introspect(
  server: McpServer,
): Promise<IntrospectionResult> {
  const [clientTransport, serverTransport] =
    InMemoryTransport.createLinkedPair();

  const client = new Client({
    name: "mcpspec-introspector",
    version: "0.1.0",
  });

  try {
    await server.connect(serverTransport);
    await client.connect(clientTransport);

    const rawCapabilities = client.getServerCapabilities() ?? {};
    const capabilities = extractCapabilities(rawCapabilities);
    const tools = rawCapabilities.tools
      ? await listAllTools(client)
      : [];
    const resources = rawCapabilities.resources
      ? await listAllResources(client)
      : [];
    const prompts = rawCapabilities.prompts
      ? await listAllPrompts(client)
      : [];

    return { tools, resources, prompts, capabilities };
  } finally {
    await client.close().catch(() => {});
    await serverTransport.close().catch(() => {});
    await clientTransport.close().catch(() => {});
  }
}

function extractCapabilities(
  raw: Record<string, unknown>,
): Record<string, Record<string, unknown> | object> {
  const result: Record<string, Record<string, unknown> | object> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (typeof value === "object" && value !== null) {
      result[key] = value as Record<string, unknown>;
    } else if (value !== undefined) {
      result[key] = {};
    }
  }
  return result;
}

async function listAllTools(
  client: Client,
): Promise<IntrospectionTool[]> {
  const tools: IntrospectionTool[] = [];
  let cursor: string | undefined;

  do {
    const response = await client.listTools(
      cursor ? { cursor } : undefined,
    );
    for (const tool of response.tools) {
      tools.push({
        name: tool.name,
        title: tool.title,
        description: tool.description,
        annotations: tool.annotations as
          | Record<string, unknown>
          | undefined,
        inputSchema: tool.inputSchema as Record<string, unknown>,
        outputSchema: tool.outputSchema as
          | Record<string, unknown>
          | undefined,
      });
    }
    cursor = response.nextCursor;
  } while (cursor);

  return tools;
}

async function listAllResources(
  client: Client,
): Promise<IntrospectionResource[]> {
  const resources: IntrospectionResource[] = [];
  let cursor: string | undefined;

  do {
    const response = await client.listResources(
      cursor ? { cursor } : undefined,
    );
    for (const resource of response.resources) {
      resources.push({
        uri: resource.uri,
        name: resource.name,
        description: resource.description,
        mimeType: resource.mimeType,
      });
    }
    cursor = response.nextCursor;
  } while (cursor);

  return resources;
}

async function listAllPrompts(
  client: Client,
): Promise<IntrospectionPrompt[]> {
  const prompts: IntrospectionPrompt[] = [];
  let cursor: string | undefined;

  do {
    const response = await client.listPrompts(
      cursor ? { cursor } : undefined,
    );
    for (const prompt of response.prompts) {
      prompts.push({
        name: prompt.name,
        description: prompt.description,
        arguments: prompt.arguments?.map((arg) => ({
          name: arg.name,
          description: arg.description,
          required: arg.required,
        })),
      });
    }
    cursor = response.nextCursor;
  } while (cursor);

  return prompts;
}

File: packages/python/src/mcpspec_dev/introspect.py
"""MCP server introspection via in-memory transport.

Connects to a FastMCP or Server instance, extracts tools, resources, prompts,
and capabilities using only MCP list endpoints. Never calls get/read/call.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from mcp.shared.memory import create_connected_server_and_client_session

from mcpspec_dev.types import (
    IntrospectionPrompt,
    IntrospectionPromptArgument,
    IntrospectionResource,
    IntrospectionResult,
    IntrospectionTool,
)

if TYPE_CHECKING:
    from mcp.client.session import ClientSession
    from mcp.server.fastmcp import FastMCP
    from mcp.server.lowlevel.server import Server
    from mcp.types import ServerCapabilities


async def introspect(
    server: FastMCP | Server[object],
) -> IntrospectionResult:
    """Introspect an MCP server and return its tools, resources, and prompts."""
    async with create_connected_server_and_client_session(
        server,
    ) as session:
        await session.initialize()

        caps = session.get_server_capabilities()
        capabilities = _extract_capabilities(caps)

        tools = await _list_all_tools(session) if caps and caps.tools else []
        resources = (
            await _list_all_resources(session) if caps and caps.resources else []
        )
        prompts = (
            await _list_all_prompts(session) if caps and caps.prompts else []
        )

        return IntrospectionResult(
            tools=tools,
            resources=resources,
            prompts=prompts,
            capabilities=capabilities,
        )


def _extract_capabilities(
    caps: ServerCapabilities | None,
) -> dict[str, Any]:
    """Convert ServerCapabilities to a plain dict, keeping only object values."""
    if caps is None:
        return {}
    result: dict[str, Any] = {}
    for key, value in caps.model_dump().items():
        if isinstance(value, dict):
            result[key] = value
        elif value is not None:
            result[key] = {}
    return result


async def _list_all_tools(
    session: ClientSession,
) -> list[IntrospectionTool]:
    """List all tools with pagination support."""
    tools: list[IntrospectionTool] = []
    cursor: str | None = None

    while True:
        response = await session.list_tools(cursor)
        for tool in response.tools:
            tools.append(
                IntrospectionTool(
                    name=tool.name,
                    title=getattr(tool, "title", None),
                    description=tool.description,
                    annotations=_annotations_to_dict(tool.annotations),
                    input_schema=dict(tool.inputSchema) if tool.inputSchema else None,
                    output_schema=getattr(tool, "outputSchema", None),
                ),
            )
        cursor = response.nextCursor
        if not cursor:
            break

    return tools


async def _list_all_resources(
    session: ClientSession,
) -> list[IntrospectionResource]:
    """List all resources with pagination support."""
    resources: list[IntrospectionResource] = []
    cursor: str | None = None

    while True:
        response = await session.list_resources(cursor)
        for resource in response.resources:
            resources.append(
                IntrospectionResource(
                    uri=str(resource.uri),
                    name=resource.name,
                    description=resource.description,
                    mime_type=resource.mimeType,
                ),
            )
        cursor = response.nextCursor
        if not cursor:
            break

    return resources


async def _list_all_prompts(
    session: ClientSession,
) -> list[IntrospectionPrompt]:
    """List all prompts with pagination support."""
    prompts: list[IntrospectionPrompt] = []
    cursor: str | None = None

    while True:
        response = await session.list_prompts(cursor)
        for prompt in response.prompts:
            arguments = None
            if prompt.arguments:
                arguments = [
                    IntrospectionPromptArgument(
                        name=arg.name,
                        description=arg.description,
                        required=arg.required,
                    )
                    for arg in prompt.arguments
                ]
            prompts.append(
                IntrospectionPrompt(
                    name=prompt.name,
                    description=prompt.description,
                    arguments=arguments,
                ),
            )
        cursor = response.nextCursor
        if not cursor:
            break

    return prompts


def _annotations_to_dict(
    annotations: object,
) -> dict[str, Any] | None:
    """Convert tool annotations to a plain dict, or None if absent."""
    if annotations is None:
        return None
    if hasattr(annotations, "model_dump"):
        data: dict[str, Any] = annotations.model_dump(exclude_none=True)
        return data if data else None
    return None
